File: execute_schema.js
#!/usr/bin/env node

/**
 * Executa schema.sql no Supabase via JavaScript
 * Usando @supabase/supabase-js library
 */

const fs = require('fs');
const path = require('path');
const https = require('https');

// Credenciais
const SUPABASE_URL = 'https://qqzgkxdgobhhivyofhuj.supabase.co';
const SERVICE_ROLE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InFxemdreGRnb2JoaGl2eW9maHVqIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3MDY1OTI4NCwiZXhwIjoyMDg2MjM1Mjg0fQ.76v6TRU-Pv5qF0GNLm5duvTAZCBklee4ndH4PiGFJ6A';

const schemaPath = path.join(__dirname, 'zona-genius-dashboard/schema.sql');

/**
 * Executa uma query SQL via Supabase SQL API
 * (usando um workaround com Supabase edge functions)
 */
async function executeSQLviaAPI(query) {
  return new Promise((resolve, reject) => {
    const payload = JSON.stringify({ query });

    const options = {
      hostname: 'qqzgkxdgobhhivyofhuj.supabase.co',
      port: 443,
      path: '/rest/v1/rpc/exec_sql',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(payload),
        'Authorization': `Bearer ${SERVICE_ROLE_KEY}`,
        'apikey': SERVICE_ROLE_KEY,
      },
    };

    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          resolve({ ok: true, status: res.statusCode, data });
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${data}`));
        }
      });
    });

    req.on('error', reject);
    req.write(payload);
    req.end();
  });
}

/**
 * Main execution
 */
async function main() {
  console.log('='.repeat(70));
  console.log('🔧 Supabase Schema Setup via JavaScript');
  console.log('='.repeat(70));

  // Ler schema
  if (!fs.existsSync(schemaPath)) {
    console.error(`❌ Arquivo não encontrado: ${schemaPath}`);
    return 1;
  }

  const schema = fs.readFileSync(schemaPath, 'utf-8');
  console.log(`\n✅ Schema carregado (${schema.length} bytes)`);

  console.log(`\n🔗 Conectando ao Supabase...`);
  console.log(`   URL: ${SUPABASE_URL}`);

  // Dividir em statements
  const statements = schema
    .split(';')
    .map(stmt => stmt.trim())
    .filter(stmt => stmt && !stmt.startsWith('--'));

  console.log(`\n📊 Total de statements: ${statements.length}`);
  console.log('\n🚀 Executando SQL via API...\n');

  let successCount = 0;
  let errorCount = 0;
  const errors = [];

  // Executar sequencialmente (não paralelo pois pode ter dependências)
  for (let i = 0; i < statements.length; i++) {
    const stmt = statements[i];
    const preview = stmt.substring(0, 70).replace(/\n/g, ' ') +
                   (stmt.length > 70 ? '...' : '');

    process.stdout.write(
      `[${String(i + 1).padStart(2, ' ')}/${statements.length}] ${
        preview.substring(0, 70).padEnd(70, ' ')
      }`
    );

    try {
      const result = await executeSQLviaAPI(stmt);
      console.log(' ✅');
      successCount++;
    } catch (error) {
      console.log(` ❌`);
      errorCount++;
      errors.push({
        num: i + 1,
        statement: stmt.substring(0, 100),
        error: error.message,
      });
    }

    // Delay de 100ms para não sobrecarregar
    await new Promise(r => setTimeout(r, 100));
  }

  // Resultado
  console.log('\n' + '='.repeat(70));
  console.log(`📊 RESULTADO FINAL:`);
  console.log(`   ✅ Sucesso: ${successCount}/${statements.length}`);
  console.log(`   ❌ Erros: ${errorCount}/${statements.length}`);
  console.log('='.repeat(70));

  if (errorCount === 0) {
    console.log('\n🎉 SUCESSO! Todas as tabelas criadas no Supabase!');
    console.log('\n📋 Tabelas criadas:');
    console.log('   ✓ organizations');
    console.log('   ✓ users');
    console.log('   ✓ assessments');
    console.log('   ✓ genius_profiles');
    console.log('   ✓ squad_recommendations');
    console.log('   ✓ genius_zone_blueprints');
    console.log('   ✓ chat_responses');
    console.log('   ✓ audit_logs');
    console.log('\n✅ RLS policies ativadas');
    console.log('✅ Índices criados');
    console.log('✅ Triggers configurados\n');
    return 0;
  } else {
    console.log(`\n⚠️  ${errorCount} statements com erro`);
    if (errors.length > 0) {
      console.log('\nPrimeiros erros:');
      errors.slice(0, 3).forEach(err => {
        console.log(`\n[${err.num}] ${err.statement}`);
        console.log(`    → ${err.error.substring(0, 120)}`);
      });
    }
    return 1;
  }
}

main().then(code => process.exit(code)).catch(err => {
  console.error('❌ Erro fatal:', err.message);
  process.exit(1);
});

File: execute_schema.py
#!/usr/bin/env python3
"""
Executa schema.sql no Supabase PostgreSQL
Conexão direta ao banco de dados
"""

import sys
import os

# Importar psycopg2
try:
    import psycopg2
    import psycopg2.extras
except ImportError:
    print("Instalando psycopg2...")
    os.system(f"{sys.executable} -m pip install --user psycopg2-binary --quiet")
    import psycopg2
    import psycopg2.extras

def main():
    # Credenciais
    config = {
        'host': 'qqzgkxdgobhhivyofhuj.supabase.co',
        'database': 'postgres',
        'user': 'postgres',
        'password': 'BBeennoorr0101',
        'port': 5432,
        'sslmode': 'require'
    }

    schema_path = 'zona-genius-dashboard/schema.sql'

    print("=" * 70)
    print("🔧 Supabase Schema Setup")
    print("=" * 70)

    # Ler schema
    if not os.path.exists(schema_path):
        print(f"❌ Arquivo não encontrado: {schema_path}")
        return 1

    with open(schema_path, 'r') as f:
        schema = f.read()

    print(f"\n✅ Schema carregado ({len(schema)} bytes)")
    print(f"\n🔗 Conectando ao Supabase...")
    print(f"   Host: {config['host']}")
    print(f"   DB: {config['database']}")
    print(f"   User: {config['user']}")

    try:
        # Conectar ao PostgreSQL
        conn = psycopg2.connect(
            host=config['host'],
            database=config['database'],
            user=config['user'],
            password=config['password'],
            port=config['port'],
            sslmode=config['sslmode'],
            connect_timeout=10
        )

        print("✅ Conectado com sucesso!\n")

        cursor = conn.cursor()

        # Dividir em statements (split por ;)
        statements = []
        current = ""
        in_string = False
        quote_char = None

        for char in schema:
            if char in ('"', "'") and (not current or current[-1] != '\\'):
                in_string = not in_string
                quote_char = char if in_string else None

            current += char

            if char == ';' and not in_string:
                stmt = current.strip()
                if stmt and not stmt.startswith('--'):
                    statements.append(stmt)
                current = ""

        # Filtrar comentários
        filtered_statements = []
        for stmt in statements:
            lines = stmt.split('\n')
            cleaned_lines = [line for line in lines if not line.strip().startswith('--')]
            cleaned = '\n'.join(cleaned_lines).strip()
            if cleaned:
                filtered_statements.append(cleaned)

        statements = filtered_statements

        print(f"📊 Total de statements: {len(statements)}\n")
        print("🚀 Executando SQL...\n")

        success_count = 0
        error_count = 0
        errors = []

        for i, statement in enumerate(statements, 1):
            preview = statement[:70].replace('\n', ' ')
            if len(statement) > 70:
                preview += "..."

            try:
                sys.stdout.write(f"[{i:2d}/{len(statements)}] {preview:<75}")
                sys.stdout.flush()

                cursor.execute(statement)
                conn.commit()

                print(" ✅")
                success_count += 1

            except psycopg2.Error as e:
                error_msg = str(e).split('\n')[0][:50]
                print(f" ❌ {error_msg}")
                error_count += 1
                errors.append({
                    'num': i,
                    'statement': statement[:100],
                    'error': str(e)
                })

            except Exception as e:
                print(f" ❌ Erro inesperado")
                error_count += 1
                errors.append({
                    'num': i,
                    'statement': statement[:100],
                    'error': str(e)
                })

        # Fechar conexão
        cursor.close()
        conn.close()

        # Resultado
        print("\n" + "=" * 70)
        print(f"📊 RESULTADO FINAL:")
        print(f"   ✅ Sucesso: {success_count}/{len(statements)}")
        print(f"   ❌ Erros: {error_count}/{len(statements)}")
        print("=" * 70)

        if error_count == 0:
            print("\n🎉 SUCESSO! Todas as tabelas criadas no Supabase!")
            print("\n📋 Tabelas criadas:")
            print("   ✓ organizations")
            print("   ✓ users")
            print("   ✓ assessments")
            print("   ✓ genius_profiles")
            print("   ✓ squad_recommendations")
            print("   ✓ genius_zone_blueprints")
            print("   ✓ chat_responses")
            print("   ✓ audit_logs")
            print("\n✅ RLS policies ativadas")
            print("✅ Índices criados")
            print("✅ Triggers configurados")
            return 0
        else:
            print(f"\n⚠️  {error_count} statements com erro")
            print("\nPrimeiros erros:")
            for err in errors[:3]:
                print(f"\n[{err['num']}] {err['statement'][:80]}")
                print(f"    → {err['error'][:120]}")
            return 1

    except psycopg2.OperationalError as e:
        print(f"❌ Erro de conexão: {e}")
        print("\n💡 Verifique:")
        print("   • Senha está correta?")
        print("   • Host está correto?")
        print("   • Supabase projeto está ativo?")
        return 1

    except Exception as e:
        print(f"❌ Erro: {e}")
        import traceback
        traceback.print_exc()
        return 1

if __name__ == '__main__':
    sys.exit(main())
